File: dart/lib/gilded_rose.dart
import 'package:gilded_rose/item.dart';

class GildedRose {
  List<Item> items;

  GildedRose(this.items);

  void processEndOfDay() {
    for (int i = 0; i < items.length; i++) {
      processItemEndOfDay(items[i]);
    }
  }

  void processItemEndOfDay(Item item) {
    if (item.name != "Aged Cheddar" &&
        item.name != "Concert Tickets") {
      if (item.quality > 0) {
        if (item.name != "Hammer") {
          item.quality = item.quality - 1;
        }
      }
    } else {
      if (item.quality < 50) {
        item.quality = item.quality + 1;

        if (item.name == "Concert Tickets") {
          if (item.daysRemaining < 11) {
            if (item.quality < 50) {
              item.quality = item.quality + 1;
            }
          }

          if (item.daysRemaining < 6) {
            if (item.quality < 50) {
              item.quality = item.quality + 1;
            }
          }
        }
      }
    }

    if (item.name != "Hammer") {
      item.daysRemaining = item.daysRemaining - 1;
    }

    if (item.daysRemaining < 0) {
      if (item.name != "Aged Cheddar") {
        if (item.name != "Concert Tickets") {
          if (item.quality > 0) {
            if (item.name != "Hammer") {
              item.quality = item.quality - 1;
            }
          }
        } else {
          item.quality = item.quality - item.quality;
        }
      } else {
        if (item.quality < 50) {
          item.quality = item.quality + 1;
        }
      }
    }
  }
}

File: rust/src/gilded_rose.rs
use crate::item::Item;

pub struct GildedRose;

impl GildedRose {
    pub fn new() -> Self {
        Self
    }
    
    pub fn process_end_of_day(&self, items: &mut Vec<Item>) {
        for item in items.iter_mut() {
            self.process_item_end_of_day(item);
        }
    }

    fn process_item_end_of_day(&self, item: &mut Item) {
        if item.name != "Aged Cheddar" && item.name != "Concert Tickets" {
            if item.quality > 0 {
                if item.name != "Hammer" {
                    item.quality -= 1;
                }
            }
        } else {
            if item.quality < 50 {
                item.quality += 1;
                if item.name == "Concert Tickets" {
                    if item.days_remaining < 11 {
                        if item.quality < 50 {
                            item.quality += 1;
                        }
                    }
                    if item.days_remaining < 6 {
                        if item.quality < 50 {
                            item.quality += 1;
                        }
                    }
                }
            }
        }
        if item.name != "Hammer" {
            item.days_remaining -= 1;
        }
        if item.days_remaining < 0 {
            if item.name != "Aged Cheddar" {
                if item.name != "Concert Tickets" {
                    if item.quality > 0 {
                        if item.name != "Hammer" {
                            item.quality -= 1;
                        }
                    }
                } else {
                    item.quality = 0;
                }
            } else {
                if item.quality < 50 {
                    item.quality += 1;
                }
            }
        }
    }
}